### PR TITLE
12520 TEv2.1 layernorm fwd triton

### PR DESCRIPTION
**Pull request by sudhu2k**
# Description

This PR introduces new functionality to the TransformerEngine, specifically providing Triton-based drop-in replacements for layernorm fwd operations. 

Fixes [12520](https://github.com/ROCm/frameworks-internal/issues/12520)
Validates [12519](https://github.com/ROCm/frameworks-internal/issues/12519)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Adds a new drop-in replacement for layernorm_fwd method.
- Adds unit tests to test out the new layernorm in test_cast.py. 
- Also creates torch reference to verify if atol=5x10^-7 is met by the hipified and triton kernels
- Adds detailed message to te_compare_results

# Checklist:

- [x ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ x] The functionality is complete
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ x] I have made corresponding changes to the documentation
- [ x] My changes generate no new warnings
- [ x] I have added tests that prove my fix is effective or that my feature works
- [x ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/triton_kernels/test_common.py [hunk 1]
```diff
@@ -145,3 +177,6 @@ def skip_mixed_16bit_float_types(in_dtype, out_dtype):
         in_dtype == torch.bfloat16 and out_dtype == torch.float16
     ):
         pytest.skip("hipified implementation does not support mixing fp16 and bf16")
+
+def IS_FP8(dtype):
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_common.py]: It duplicates transformer_engine.pytorch.triton_kernels.common.is_fp8_torch_dtype()

File: tests/pytorch/triton_kernels/test_common.py [hunk 2]
```diff
@@ -55,6 +55,8 @@ def te_compare_results(t, r, atol, rtol, msg):
     assert atol > 0, "Absolute tolerance must be positive."
     assert rtol > 0, "Relative tolerance must be positive."
     dtype = t.dtype
+    t_orig = t
+    r_orig = r
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_common.py]: Looks unused

File: tests/pytorch/triton_kernels/test_common.py [hunk 3]
```diff
@@ -64,6 +66,21 @@ def te_compare_results(t, r, atol, rtol, msg):
     rtol_mismatch[nonzero_r] = torch.abs(diff[nonzero_r] / r[nonzero_r]) > rtol
     mismatch = atol_mismatch & (~nonzero_r | rtol_mismatch)
     has_mismatch = torch.any(mismatch).item()
+
+    max_rel_diff = 0.0 # Default to 0.0 if no non-zero reference values
+    max_abs_diff = 0.0 
+    max_abs_diff_indices = None
+    max_rel_diff_indices = None
+    if has_mismatch:
+        max_abs_diff = torch.max(torch.abs(diff[mismatch])).item()
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_common.py]: To get max diff and value there is no need to use diff[mismatch], or create abs_diff tensor, diff can be used directly.
For rel diff rather rtol_mismatch should be used because mismatch may have indices of r==0.

File: tests/pytorch/triton_kernels/test_layernorm.py [hunk 4]
```diff
@@ -20,8 +22,10 @@
 )
 from transformer_engine.pytorch.triton_kernels.layernorm import (
     te_layernorm_bwd_triton,
+    te_layernorm_fwd_triton,
 )
 from test_common import (
+    IS_FP8,
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_layernorm.py]: is_fp8_torch_dtype is imported eralier

File: tests/pytorch/triton_kernels/test_layernorm.py [hunk 5]
```diff
@@ -49,12 +83,14 @@
 ]
 
 all_boolean = [False, True]
+torch.manual_seed(42)
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_layernorm.py]: Why is it needed?

File: tests/pytorch/triton_kernels/test_layernorm.py [hunk 6]
```diff
@@ -72,25 +108,148 @@ def test_layernorm_fwd_bwd_triton(in_dtype, out_dtype, M, N, zero_centered_gamma
     epsilon = 1e-5
 
     # Run Hipified forward reference.
-    y_hipified = torch.empty((M, N), dtype=out_dtype, device="cuda")
+    quantizer_hipified = None
+    quantizer_triton = None
+    scale_ref, amax_ref = torch.tensor(1.0), torch.zeros(1, dtype=torch.float32, device='cuda')
+    if is_fp8_torch_dtype(out_dtype):
+        scale_triton = torch.rand(1, dtype=torch.float32, device='cuda') * 3.0 - 2.0
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_layernorm.py:113]: Is it to make specific values range?

File: tests/pytorch/triton_kernels/test_layernorm.py [hunk 7]
```diff
@@ -72,25 +108,148 @@ def test_layernorm_fwd_bwd_triton(in_dtype, out_dtype, M, N, zero_centered_gamma
...
+        y_hipified,
+        y_ref,
+        atol_fwd,
+        rtol_fwd,
+        lambda msg: f"y does not match hip <-> ref\n\n{msg}\n",
+    )
+
+    compare_results(
+        "te",
+        y_triton,
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_layernorm.py:193]: triton<->hip is compared above

File: tests/pytorch/triton_kernels/test_layernorm.py [hunk 8]
```diff
@@ -72,25 +108,148 @@ def test_layernorm_fwd_bwd_triton(in_dtype, out_dtype, M, N, zero_centered_gamma
...
+            y_hipified_transpose,
+            atol_fwd,
+            rtol_fwd,
+            lambda msg: f"y transpose does not match triton <-> hip\n\n{msg}\n",
+        )
+    atol_stats, _ = get_tolerances(torch.float32)
+    rtol_stats = 5e-5
+
+    compare_results(
+        fwd_cmp,
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_layernorm.py:212]: why not use "te" like in other compare_results?

File: transformer_engine/pytorch/ops/basic/layer_norm.py [hunk 9]
```diff
@@ -17,7 +17,7 @@
 from transformer_engine_torch import layernorm_bwd, layernorm_fwd
 from torch.utils.cpp_extension import IS_HIP_EXTENSION
 if IS_HIP_EXTENSION:
-    # TODO: bring back te_layernorm_fwd_triton after refactoring
+    from ...triton_kernels.layernorm import te_layernorm_fwd_triton
```
Comment (ipanfilo) [inline transformer_engine/pytorch/ops/basic/layer_norm.py]: It should be merged with the next line

**Review by ipanfilo — CHANGES_REQUESTED** [review]

File: tests/pytorch/triton_kernels/test_layernorm.py (hunk 5 above)
Comment (sudhu2k) [reply tests/pytorch/triton_kernels/test_layernorm.py]: It is just to validate results while I was debugging, making sure I was getting the same values every time I run the tests. I can remove it now.

File: tests/pytorch/triton_kernels/test_common.py (hunk 3 above)
Comment (sudhu2k) [reply tests/pytorch/triton_kernels/test_common.py]: Right, it makes sense for max_abs_diff. But shouldn't we use nonzero_r mask for max_rel_diff ? It'll mask away the indices of r==0.

File: tests/pytorch/triton_kernels/test_layernorm.py (hunk 4 above)
Comment (sudhu2k) [reply tests/pytorch/triton_kernels/test_layernorm.py]: I will fix it!

File: tests/pytorch/triton_kernels/test_layernorm.py (hunk 6 above)
Comment (sudhu2k) [reply tests/pytorch/triton_kernels/test_layernorm.py:113]: Yes, similar to the cpp tests.
https://github.com/ROCm/TransformerEngine/blob/cc960418370fd928e840ce6eeec28b6c9b848470/tests/cpp/test_common.cu#L781-L785

File: tests/pytorch/triton_kernels/test_layernorm.py (hunk 7 above)
Comment (sudhu2k) [reply tests/pytorch/triton_kernels/test_layernorm.py:193]: Will fix it

File: tests/pytorch/triton_kernels/test_layernorm.py (hunk 8 above)
Comment (sudhu2k) [reply tests/pytorch/triton_kernels/test_layernorm.py:212]: fwd_cmp is actually assigned to "te", so I'll use fwd_cmp for other compare_results as well.

File: tests/pytorch/triton_kernels/test_common.py (hunk 3 above)
Comment (sudhu2k) [reply tests/pytorch/triton_kernels/test_common.py]: Also, since there's a mismatch only when both the atol_mismatch and rtol_mismatch is True, doesn't it help us to just check the mismatched rows and columns?

Comment (ipanfilo) [reply tests/pytorch/triton_kernels/test_common.py]: mismatch is True if atol_mismatch and (rtol_mismatch OR r==0)

File: transformer_engine/pytorch/triton_kernels/layernorm.py [hunk 10]
```diff
@@ -445,7 +454,109 @@ def _layernorm_bwd_dwdb_triton_v2(
...
+
+    # For MXFP8, we do regular layernorm and then quantize it separately
+    if isinstance(quantizer, MXFP8Quantizer):
+        use_cast_transpose_triton =  bool( int(os.environ.get('NVTE_USE_CAST_TRANSPOSE_TRITON', '0')) )
+        quantize_func = te_quantize_triton if use_cast_transpose_triton else tex.quantize
+        ln_out = quantize_func(ln_out)
+    
+    # Reduce and find amax if "not APPLY_ATOMIC" is True.
+    if IS_FP8 and not APPLY_ATOMIC:
+        _layernorm_fwd_reduce_triton[(triton.cdiv(M, 256),)](
```
Comment (Micky774) [inline transformer_engine/pytorch/triton_kernels/layernorm.py:546]: iirc `triton.cdiv` can only be called inside kernels -- is this code-path tested?

Comment (sudhu2k) [reply transformer_engine/pytorch/triton_kernels/layernorm.py:546]: triton.cdiv can be called outside, only triton language specific (tl) functionalities can be used inside kernels and those are optimized to run on a GPU, like tl.max, tl.sum, there is a tl.cdiv as well which you're referring to and can only be used inside a kernel. There are also several triton functions that can be used outside like triton.reinterpret(). Yes, it works well on my unit tests.

File: tests/pytorch/triton_kernels/test_common.py [hunk 11]
```diff
@@ -64,6 +64,19 @@ def te_compare_results(t, r, atol, rtol, msg):
     rtol_mismatch[nonzero_r] = torch.abs(diff[nonzero_r] / r[nonzero_r]) > rtol
     mismatch = atol_mismatch & (~nonzero_r | rtol_mismatch)
     has_mismatch = torch.any(mismatch).item()
+
+    max_rel_diff = 0.0 # Default to 0.0 if no non-zero reference values
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_common.py:66]: It seems that if you calculate max_rel_diff from rel_diff, it should not be a problem if there is no non-zero reference values. And it also avoid duplicated calculation of diff[nonzero_r] / r[nonzero_r]

File: tests/pytorch/triton_kernels/test_common.py [hunk 12]
```diff
@@ -144,4 +172,4 @@ def skip_mixed_16bit_float_types(in_dtype, out_dtype):
     if (in_dtype == torch.float16 and out_dtype == torch.bfloat16) or (
         in_dtype == torch.bfloat16 and out_dtype == torch.float16
     ):
-        pytest.skip("hipified implementation does not support mixing fp16 and bf16")
+        pytest.skip("hipified implementation does not support mixing fp16 and bf16")
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_common.py]: No trailing EOL - lint error

**Review by ipanfilo — APPROVED** [review]

**Review by ipanfilo — APPROVED** [review]

File: tests/pytorch/triton_kernels/test_layernorm.py [hunk 13]
```diff
@@ -72,25 +106,148 @@ def test_layernorm_fwd_bwd_triton(in_dtype, out_dtype, M, N, zero_centered_gamma
...
+        otype=torch_dtype_to_te_dtype(out_dtype),
+        sm_margin=get_fwd_ln_sm_margin(),
+        zero_centered_gamma=zero_centered_gamma
+    )
+
+    # Run forward
+    fwd_cmp = "te"
+    atol_fwd, rtol_fwd = get_tolerances(out_dtype)
+    
+    # Uncommenting this creates some fp32 test failures
```
Comment (wenchenvincent) [inline tests/pytorch/triton_kernels/test_layernorm.py:155]: Is there still a failure?

Comment (sudhu2k) [reply tests/pytorch/triton_kernels/test_layernorm.py:155]: Yes, there is. The hipified kernels too fail with tighter tolerances against my reference.

**Review by wenchenvincent — APPROVED** [review]
LGTM.
